sonar-core/src/main/java/org/sonar/core/util/UuidGeneratorImpl.java

```diff
@@ -0,0 +1,113 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+public final class UuidGeneratorImpl implements UuidGenerator {
+
+  private final FullNewUuidGenerator fullNewUuidGenerator = new FullNewUuidGenerator();
+
+  @Override
+  public byte[] generate() {
+    return fullNewUuidGenerator.get();
+  }
+
+  @Override
+  public WithFixedBase withFixedBase() {
+    return new FixedBasedUuidGenerator();
+  }
+
+  private static class UuidGeneratorBase {
+    private final byte[] buffer = new byte[15];
+    // We only use bottom 3 bytes for the sequence number. Paranoia: init with random int so that if JVM/OS/machine goes down, clock slips
+    // backwards, and JVM comes back up, we are less likely to be on the same sequenceNumber at the same time:
+    private final AtomicInteger sequenceNumber = new AtomicInteger(new SecureRandom().nextInt());
+    private final byte[] secureMungedAddress = MacAddressProvider.getSecureMungedAddress();
+    // Used to ensure clock moves forward
+    private long lastTimestamp = 0L;
+
+    void initBase(int sequenceId) {
+      long timestamp = System.currentTimeMillis();
+
+      synchronized (this) {
+        // Don't let timestamp go backwards, at least "on our watch" (while this JVM is running). We are still vulnerable if we are
+        // shut down, clock goes backwards, and we restart... for this we randomize the sequenceNumber on init to decrease chance of
+        // collision:
+        timestamp = Math.max(lastTimestamp, timestamp);
+
+        if (sequenceId == 0) {
+          // Always force the clock to increment whenever sequence number is 0, in case we have a long time-slip backwards:
+          timestamp++;
+        }
+
+        lastTimestamp = timestamp;
+      }
+
+      // Only use lower 6 bytes of the timestamp (this will suffice beyond the year 10000):
+      putLong(buffer, timestamp, 0, 6);
+
+      // MAC address adds 6 bytes:
+      System.arraycopy(secureMungedAddress, 0, buffer, 6, secureMungedAddress.length);
+    }
+
+    protected byte[] generate(int increment) {
+      // Sequence number adds 3 bytes
+      putLong(buffer, increment, 12, 3);
+
+      return buffer;
+    }
+
+    int getSequenceId() {
+      return sequenceNumber.incrementAndGet() & 0xffffff;
+    }
+
+    /** Puts the lower numberOfLongBytes from l into the array, starting index pos. */
+    private static void putLong(byte[] array, long l, int pos, int numberOfLongBytes) {
+      for (int i = 0; i < numberOfLongBytes; ++i) {
+        array[pos + numberOfLongBytes - i - 1] = (byte) (l >>> (i * 8));
+      }
+    }
+  }
+
+  private final class FullNewUuidGenerator extends UuidGeneratorBase implements Supplier<byte[]> {
+
+    @Override
+    public byte[] get() {
+      int sequenceId = getSequenceId();
+      initBase(sequenceId);
+      return super.generate(sequenceId);
+    }
+  }
+
+  private class FixedBasedUuidGenerator extends UuidGeneratorBase implements WithFixedBase {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2694)